plugin/resolver/source_remote.go

```diff
@@ -0,0 +1,361 @@
+	// Useful for debugging and statistics on the registry side.
+	UserAgent string
+}
+
+// regVersion represents a version of a plugin in the registry
+type regVersion struct {
+	Version   Version       `json:"version"`
+	Platforms []regPlatform `json:"platforms"`
+}
+
+// regPlatform represents an available platform for a plugin version in the registry
+type regPlatform struct {
+	OS   string `json:"os"`
+	Arch string `json:"arch"`
+}
+
+// regDownloadInfo represents the download info for a specific platform of a plugin version in the registry
+type regDownloadInfo struct {
+	OS          string `json:"os"`
+	Arch        string `json:"arch"`
+	DownloadURL string `json:"download_url"`
+}
+
+// regError represents an error response.
+type regError struct {
+	Code    string `json:"code"`
+	Message string `json:"message"`
+}
+
+// registryError implements the error interface.
+func (err regError) Error() string {
+	return fmt.Sprintf("[code=%s]: %s", err.Code, err.Message)
+}
+
+// Lookup returns the versions found of the plugin with the given name.
+func (source RemoteSource) Lookup(ctx context.Context, name Name) ([]Version, error) {
+	versions, err := source.fetchVersions(ctx, name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup plugin versions in the registry: %w", err)
+	}
+	var matches []Version
+	for _, version := range versions {
+		hasPlatform := slices.ContainsFunc(version.Platforms, func(platform regPlatform) bool {
+			return platform.OS == runtime.GOOS && platform.Arch == runtime.GOARCH
+		})
+		if hasPlatform {
+			matches = append(matches, version.Version)
+		}
+	}
+	return matches, nil
+}
+
+// call makes a http request to the registry with the given timeout.
+func (source RemoteSource) call(req *http.Request, timeout time.Duration) (*http.Response, error) {
+	if source.UserAgent != "" {
+		req.Header.Set("User-Agent", source.UserAgent)
+	}
+	client := &http.Client{
+		Timeout: timeout,
+	}
+	return client.Do(req)
+}
+
+// decodeBody decodes the http response from the registry into the provided value.
+func (source RemoteSource) decodeBody(resp *http.Response, v interface{}) error {
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		var errResp struct {
+			Error regError `json:"error"`
+		}
+		if err := json.NewDecoder(resp.Body).Decode(&errResp); err != nil {
+			return fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+		}
+		return errResp.Error
+	}
+	return json.NewDecoder(resp.Body).Decode(v)
+}
+
+// fetchVersions looks up the plugin versions in the registry.
+func (source RemoteSource) fetchVersions(ctx context.Context, name Name) ([]regVersion, error) {
+	url := fmt.Sprintf("%s/v1/plugins/%s/%s/versions", source.BaseURL, name.Namespace(), name.Short())
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Accept", "application/json")
+	resp, err := source.call(req, 10*time.Second)
```

Comment:
Perhaps we should put this `10*time.Second` into a constant to have a single place to control default http request timeout?

Reply:
I will make a const for it and also download timeout. It's not something that we will want to modify but just in case we add more endpoints for the registry it might be useful to have a const ready for it.